Issue #1502 add a name field in GenesisClassJson (Unfinished)

Yogalholic opened the pull request:
resolves #1502 
resolves DOJ-130

kariy reviewed (commented):
First of all, thanks for the contribution! 

I believe we can do this with a simpler approach. During [this phase](https://github.com/dojoengine/dojo/blob/504b83e045202ee52d19404f644c0623d98f05c3/crates/katana/primitives/src/genesis/json.rs#L268-L322), where we compute the class hashes of all the classes, at the same time we can also build up a mapping of class name to its respective class hash. This mapping can later then be used throughout the method for resolving any class names to its class hash.

glihm commented:
@Yogalholic how are you feeling about this? Do you need any help / additional information?

Yogalholic commented:
As you can see in me last PR, I am  struggling with adding new GenesisJsonError and how to handle FieldElement.

kariy commented:
> As you can see in me last PR, I am struggling with adding new GenesisJsonError and how to handle FieldElement.

What trouble are you having? Can you point out where in the code that you have trouble with?

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -99,6 +99,15 @@ pub struct GenesisClassJson {
     /// The class hash of the contract. If not provided, the class hash is computed from the
     /// class at `path`.
     pub class_hash: Option<ClassHash>,
+    #[serde(default, skip_serializing)] 
+    name: Option<String>,
```
Why did you choose to ignore this field during serialization? The name should remain if we write a genesis file I suppose.

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -249,8 +266,15 @@ impl GenesisJson {
     pub fn resolve_class_artifacts(
         &mut self,
         base_path: impl AsRef<Path>,
+        class_name_map: &HashMap<String, ClassHash>
```
```suggestion
```

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -249,8 +266,15 @@ impl GenesisJson {
+        class_name_map: &HashMap<String, ClassHash>
     ) -> Result<(), GenesisJsonError> {
         for entry in &mut self.classes {
+             if let Some(name) = &entry.name {
+                if let Some(hash) = class_name_map.get(name) {
+                entry.class_hash = Some(*hash);
+                continue;  
+                }   
+            } 
```
```suggestion
```

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -234,8 +244,15 @@ impl GenesisJson {
         path.pop();
 
         let mut genesis: Self = serde_json::from_reader(BufReader::new(file))?;
+        let mut class_name_map = HashMap::new();
+         for class in &genesis.classes {
+            if let Some(name) = &class.name {
+                class_name_map.insert(name.to_string(), class.class_hash.unwrap());  
+            }
+         }
```
```suggestion
```
May not be used at all at this point, see comment below.

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -234,8 +244,15 @@ impl GenesisJson {
+
         // resolves the class paths, if any
-        genesis.resolve_class_artifacts(path)?;
+        genesis.resolve_class_artifacts(path, &class_name_map)?;
```
```suggestion
        genesis.resolve_class_artifacts(path)?;
```

This functions resolves the artifacts, which consist on loading the class JSON into memory to be available for genesis operation. The name here shouldn't be relevant as we are expecting to use the names into other entries, not the classes themselves.

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -265,81 +289,97 @@ impl TryFrom<GenesisJson> for Genesis {
+                    .into_par_iter()
+                    .map(|entry| {
+                        let GenesisClassJson { class, class_hash, name } = entry;
+                        // Use name if present
+                        if let (Some(name), Some(class_hash)) = (name, class_hash) {
+                    name_to_class_hash.insert(name, class_hash);
+                }
```
```suggestion
```
I think this may be removed, in favor of the code at the end of this function. If the name is here, we insert the mapping. Because at the end of this function, a `class_hash` will always be defined (from the file, or from the artifacts).

glihm reviewed (commented):
Thanks for the work being done here.
Some first comments.

Also, do not hesitate to run the `scripts/rust_fmt.sh` to ensure you've the correct formatting. It helps for the review. :)

glihm commented:
Hey @Yogalholic, do you need any support or info to tackle this? Don't hesitate to reach out if yes. :+1:

Yogalholic commented:
In my last PR I have implemented your suggestions. I still don't know how to add a new GenesisJsonError

glihm commented:
> In my last PR I have implemented your suggestions. I still don't know how to add a new GenesisJsonError

This one you mean? https://github.com/dojoengine/dojo/blob/f4c8111aefe2672d04356df64914871a5bcb536d/crates/katana/primitives/src/genesis/json.rs#L153

Yogalholic commented:
Any feedback please? I have run the rust_fmt script

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -312,34 +320,99 @@ impl TryFrom<GenesisJson> for Genesis {
             })
             .collect::<Result<_, GenesisJsonError>>()?;
 
+        // Populate the classes and name_to_class_hash
```
```suggestion
```

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -312,34 +320,99 @@ impl TryFrom<GenesisJson> for Genesis {
+                Some(NameOrHash::ClassHash(class_hash)) => class_hash,
+                Some(NameOrHash::ClassName(class_name)) => *name_to_class_hash
+                    .get(&class_name)
+                    .ok_or_else(|| value_out_of_range_error::ValueOutOfRangeError)?,
```
Here we want to return a new error something like:
```rust
#[error("Class name not found in the genesis classes: {0}")]
InvalidClassName(String)
```
You should add this to the `GenesisJsonError`.

This applies to all conversions from name that fails.

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -312,34 +320,99 @@ impl TryFrom<GenesisJson> for Genesis {
+            }
+        };
+
+        // Populate the classes and name_to_class_hash
```
Why do we have this code two times? Did you mean to replace it?

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -530,8 +629,30 @@ pub fn to_base64(genesis: GenesisJson) -> Result<Vec<u8>, GenesisJsonError> {
+    // Deserialize JSON
+    let mut genesis_json: GenesisJson = serde_json::from_slice(&decoded)?;
+
+    // Populate name field
```
What do you want to achieve here? The name of the class is supposed to be given by the user into the genesis json file. :) We don't need to retrieve it from the artifact file itself.

Yogalholic replied:
OK. I have removed the for loop

glihm commented on crates/katana/primitives/src/genesis/json.rs:
```diff
@@ -312,34 +320,99 @@ impl TryFrom<GenesisJson> for Genesis {
                 };
 
+                // Add the class name and class hash to the mapping
+                let _ = name.map(|name| {
```
We don't want the user to define 2 classes with the same name. Could we add a check to ensure that the names are unique for only ONE class?

glihm reviewed (changes requested):
Thanks for the progress here @Yogalholic!
I've left some comment. Also, if you have more doubt don't hesitate to reach out on the discord we would be happy to help!

You can also start to add the name for testing once you're confident with the changes. :+1:

glihm commented:
> I have run the rust_fmt script

Thanks for that. You could also run the `clippy` script to prepare your code for the CI. :+1:

Yogalholic commented:
I have implemented every suggestion, the code is compiling fine.  I updated the first unit test (deserialize_from_json() )  and modified the test-genesis.json to replace the "0x08" classhash with the random name "MyErc20". The test is successful and now I will update the rest of them.

glihm commented:
> I have implemented every suggestion, the code is compiling fine. I updated the first unit test (deserialize_from_json() ) and modified the test-genesis.json to replace the "0x08" classhash with the random name "MyErc20". The test is successful and now I will update the rest of them.

Awesome, thanks for the heads up @Yogalholic and the work done here. A recent work was merged on Katana for the allocations to compute the total supply automatically for the fee token. You have a rebase here. Don't hesitate to ping us if you have any question/difficulty while merging.

glihm commented:
@Yogalholic don't forget to rebase on main to ensure tests can be run on the CI. 👍

Yogalholic commented:
I have finished working on the issue, the code compiles and tests are successful but clippy is failing
